packages/test/test-end-to-end-tests/src/test/gc/gcUnreferencedTimestamp.spec.ts

```diff
@@ -132,165 +132,461 @@ describeFullCompat("GC unreferenced timestamp", (getTestObjectProvider) => {
+         * references the node which was unreferenced in previous summary.
+         */
+        it(`Scenario 7 - Two new nodes B and C transitively add reference to an unreferenced node D, B and C ` +
+            `become referenced and remove the reference to D`, async () => {
+            const summarizerClient = await getNewSummarizer();
+
+            // Create data store D and mark it referenced by storing its handle in data store A.
+            const dataStoreD = await dataObjectFactory.createInstance(dataStoreA.containerRuntime);
+            dataStoreA._root.set("dataStoreD", dataStoreD.handle);
+            await provider.ensureSynchronized();
+
+            // Remove the reference to D which marks it as unreferenced.
+            dataStoreA._root.delete("dataStoreD");
+
+            // 1. Get summary 1 and validate that D is has unreferenced timestamp. E = [].
+            const timestamps1 = await getUnreferencedTimestamps(summarizerClient);
+            const dsDTime1 = timestamps1.get(dataStoreD.id);
+            assert(dsDTime1 !== undefined, `D should have unreferenced timestamp`);
+
+            // 2. Create data stores B and C. E = [].
+            const dataStoreB = await dataObjectFactory.createInstance(dataStoreA.containerRuntime);
+            const dataStoreC = await dataObjectFactory.createInstance(dataStoreA.containerRuntime);
+
+            // 3. Add reference from B to C. E = [].
+            dataStoreB._root.set("dataStoreC", dataStoreC.handle);
+
+            // 4. Add reference from C to D. E = [].
+            dataStoreC._root.set("dataStoreD", dataStoreD.handle);
+
+            // 5. Add reference from A to B. E = [A -> B, B -> C, C -> D].
+            dataStoreA._root.set("dataStoreB", dataStoreB.handle);
+            await provider.ensureSynchronized();
+
+            // 6. Remove reference from C to D. E = [A -> B, B -> C].
+            dataStoreC._root.delete("dataStoreD");
+
+            // 7. Get summary 2 and validate that D's unreferenced timestamps updated. E = [A -> B, B -> C].
+            const timestamps2 = await getUnreferencedTimestamps(summarizerClient);
+            const dsDTime2 = timestamps2.get(dataStoreD.id);
+            assertTestFails(dsDTime2 !== undefined && dsDTime2 > dsDTime1, `D's timestamp should have updated`);
+        }).timeout(20000);
+
+        /**
+         * Validates that references added by unreferences nodes do not show up as references.
+         * 1. Summary 1 at t1. V = [A*, B, C]. E = []. B and C have unreferenced time t1.
+         * 2. Op adds reference from B to C. E = [B -> C].
+         * 3. Summary 2 at t2. V = [A*, B, C]. E = [B -> C]. B and C have unreferenced time t1.
+         * Validates that the unreferenced time for B and C is still t1.
+         */
+         it(`Scenario 8 - An unreferenced node B adds reference to another node C`, async () => {
```

Comment:
Is there a chance to reuse some code between these scenarios?

Reply:
I don't think so. I moved out the common parts already. Maybe I can squeeze some more things together but it would make the tests harder to read.